src/ck-core/ckarray.C

```diff
@@ -1539,6 +1548,12 @@ void CkArray::recvBroadcast(CkMessage* m) {
     bool doFree = true; // free it since all ops are done
     broadcaster->deliver(msg, (ArrayElement*)localElemVec[0], doFree);
   } else if (zc_msgtype == CMK_ZC_BCAST_RECV_MSG && len > 0 ) {
+
+    // Set Array manager for this message as it is required for ZC Post API
+    env->setMsgtype(ForArrayEltMsg);
```

Comment:
Won't this cause a race? I think once the message is received we should not be modifying the message? Same goes for a few things below. I think we can use the group id in the envelope rather than explicitly setting and getting array manager (they should be the same).

Reply:
The way the Post API is implemented for chare arrays is that a `CMK_ZC_BCAST_RECV_MSG` reaches only one array element of one of the PEs of the node. So, at this point, it is the only PE that is operating on this message. For that reason, it doesn't cause a race condition. 

However, I like the idea of not modifying the message type and getting the array manager using the group id. I'll incorporate that change.